### PR TITLE
Grant LB policies write access to connection stream info

### PR DESCRIPTION
**Pull request by yanavlasov**
Additional Description:
We have uses cases where LB policies need to be able to write to request metadata when choosing or picking hosts. The metadata can be used in logging or by HTTP filters.

This PR makes access to the request stream info non `const`.

Risk Level: Low
Testing: Unit tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: source/extensions/clusters/original_dst/original_dst_cluster.cc [hunk 1]
```diff
@@ -105,7 +105,7 @@ HostConstSharedPtr OriginalDstCluster::LoadBalancer::chooseHost(LoadBalancerCont
 Network::Address::InstanceConstSharedPtr
 OriginalDstCluster::LoadBalancer::filterStateOverrideHost(LoadBalancerContext* context) {
   const auto streamInfos = {
-      context->requestStreamInfo(),
+      static_cast<const StreamInfo::StreamInfo*>(context->requestStreamInfo()),
```
Comment (wbpcode) [inline source/extensions/clusters/original_dst/original_dst_cluster.cc:108]: I think we should use const_cast for this case.

File: source/extensions/clusters/original_dst/original_dst_cluster.cc [hunk 2]
```diff
@@ -154,7 +154,7 @@ OriginalDstCluster::LoadBalancer::metadataOverrideHost(LoadBalancerContext* cont
     return nullptr;
   }
   const auto streamInfos = {
-      context->requestStreamInfo(),
+      static_cast<const StreamInfo::StreamInfo*>(context->requestStreamInfo()),
```
Comment (wbpcode) [inline source/extensions/clusters/original_dst/original_dst_cluster.cc:157]: ditto.

**Review by wbpcode — COMMENTED** [review]
LGTM overall with only two comments.

File: test/integration/load_balancers/custom_lb_policy.h [hunk 3]
```diff
@@ -24,7 +24,12 @@ class ThreadAwareLbImpl : public Upstream::ThreadAwareLoadBalancer {
   public:
     LbImpl(const Upstream::HostSharedPtr& host) : host_(host) {}
 
-    Upstream::HostConstSharedPtr chooseHost(Upstream::LoadBalancerContext*) override {
+    Upstream::HostConstSharedPtr chooseHost(Upstream::LoadBalancerContext* context) override {
+      if (context && context->requestStreamInfo()) {
```
Comment (nezdolik) [inline test/integration/load_balancers/custom_lb_policy.h:28]: should dynamic metadata be populated in this method here or rather at a call site? Does it need to be populated by default for all callers/in all tests that pass non null context?

File: source/extensions/clusters/original_dst/original_dst_cluster.cc (hunk 1 above)
Comment (yanavlasov) [reply source/extensions/clusters/original_dst/original_dst_cluster.cc]: Done

File: source/extensions/clusters/original_dst/original_dst_cluster.cc (hunk 2 above)
Comment (yanavlasov) [reply source/extensions/clusters/original_dst/original_dst_cluster.cc]: Done

File: test/integration/load_balancers/custom_lb_policy.h (hunk 3 above)
Comment (yanavlasov) [reply test/integration/load_balancers/custom_lb_policy.h]: I see how it can be confusing. I have created a dedicated test LB that writes into metadata in the `test/common/upstream/cluster_manager_impl_test.cc` test suite and reverted changes to the one in the test/integration

**Review by nezdolik — APPROVED** [review]
